src/main/c/CMakeLists.txt

```diff
@@ -59,62 +60,85 @@ set(SEASOCKS_SOURCE_FILES
         util/Json.cpp
         util/PathHandler.cpp
         util/RootPageHandler.cpp
+    )
+    if (DEFLATE_SUPPORT)
+        target_sources(${NAME} PRIVATE seasocks/ZlibContext.cpp)
+    else ()
+        target_sources(${NAME} PRIVATE seasocks/ZlibContextDisabled.cpp)
+    endif ()
+    target_include_directories(${NAME}
+        PUBLIC
+            $<TARGET_PROPERTY:seasocks_headers,INTERFACE_INCLUDE_DIRECTORIES>
+    )
+    if (DEFLATE_SUPPORT)
+        target_include_directories(${NAME}
+            PRIVATE
+                $<BUILD_INTERFACE:$<TARGET_PROPERTY:ZLIB::ZLIB,INTERFACE_INCLUDE_DIRECTORIES>>
         )
+    endif ()
+endfunction()
 
-if (DEFLATE_SUPPORT)
-    set(SEASOCKS_SOURCE_FILES ${SEASOCKS_SOURCE_FILES} seasocks/ZlibContext.cpp)
-else()
-    set(SEASOCKS_SOURCE_FILES ${SEASOCKS_SOURCE_FILES} seasocks/ZlibContextDisabled.cpp)
-endif()
-
-add_library(seasocks_obj OBJECT ${SEASOCKS_SOURCE_FILES})
-target_include_directories(seasocks_obj PUBLIC
-    $<INSTALL_INTERFACE:${CMAKE_INSTALL_INCLUDEDIR}/>
-    $<BUILD_INTERFACE:${CMAKE_CURRENT_SOURCE_DIR}/.>
+add_library(seasocks_headers INTERFACE)
+target_include_directories(seasocks_headers
+    INTERFACE
+        $<INSTALL_INTERFACE:${CMAKE_INSTALL_INCLUDEDIR}/.>
+        $<BUILD_INTERFACE:${CMAKE_CURRENT_SOURCE_DIR}/.>
+        $<BUILD_INTERFACE:${PROJECT_BINARY_DIR}/.>
 )
-set_property(TARGET seasocks_obj PROPERTY POSITION_INDEPENDENT_CODE TRUE)
 
-add_library(seasocks STATIC $<TARGET_OBJECTS:seasocks_obj> $<TARGET_OBJECTS:embedded>)
-add_library(Seasocks::seasocks ALIAS seasocks)
-target_include_directories(seasocks PUBLIC
-    $<INSTALL_INTERFACE:${CMAKE_INSTALL_INCLUDEDIR}/>
-    $<BUILD_INTERFACE:${CMAKE_CURRENT_SOURCE_DIR}/.>
-)
-target_link_libraries(seasocks PRIVATE ${CMAKE_THREAD_LIBS_INIT})
-if (DEFLATE_SUPPORT)
-    target_link_libraries(seasocks PRIVATE "${ZLIB_LIBRARIES}")
-endif()
+if (SEASOCKS_SHARED OR CMAKE_POSITION_INDEPENDENT_CODE)
+    add_seasocks_obj_library(seasocks_pic_obj)
+    set_property(TARGET seasocks_pic_obj PROPERTY POSITION_INDEPENDENT_CODE TRUE)
+endif ()
 
-add_library(seasocks_so SHARED $<TARGET_OBJECTS:seasocks_obj> $<TARGET_OBJECTS:embedded>)
-add_library(Seasocks::seasocks_so ALIAS seasocks_so)
-target_include_directories(seasocks_so PUBLIC ${ZLIB_INCLUDE_DIRS} 
-    $<INSTALL_INTERFACE:${CMAKE_INSTALL_INCLUDEDIR}/>
-    $<BUILD_INTERFACE:${CMAKE_CURRENT_SOURCE_DIR}/.>
-)
-if (DEFLATE_SUPPORT)
-    target_link_libraries(seasocks_so PRIVATE ${CMAKE_THREAD_LIBS_INIT} "${ZLIB_LIBRARIES}")
-endif()
-set_target_properties(seasocks_so PROPERTIES OUTPUT_NAME seasocks VERSION ${PROJECT_VERSION})
+if (SEASOCKS_SHARED)
```

Comment:
Since the main difference between the static / shared part is target name, they can be extracted into a macro I guess. This would reduce the code duplication.

Reply:
Agreed.
I didn't because I wasn't sure you would agree with the complexity I was adding.